allennlp/data/dataset_readers/penn_tree_bank.py

```diff
@@ -0,0 +1,173 @@
+                respect to a gold parse tree. If a span is not contained
+                within the tree, a span will have a ``NO-LABEL`` label.
+        """
+        # pylint: disable=arguments-differ
+        text_field = TextField([Token(x) for x in tokens], token_indexers=self._token_indexers)
+        pos_tag_field = SequenceLabelField(pos_tags, text_field, "pos_tags")
+        fields = {"tokens": text_field, "pos_tags": pos_tag_field}
+        spans: List[Field] = []
+        gold_labels = []
+
+        if gold_tree is not None:
+            gold_spans_with_pos_tags = self._get_gold_spans(gold_tree, 0, OrderedDict()).items()
+            gold_spans = {span: label for (span, label)
+                          in gold_spans_with_pos_tags if "-POS" not in label}
+        else:
+            gold_spans = None
+        for start, end in enumerate_spans(tokens):
+            spans.append(SpanField(start, end, text_field))
+
+            if gold_spans is not None:
+                if (start, end) in gold_spans.keys():
+                    gold_labels.append(gold_spans[(start, end)])
+                else:
+                    gold_labels.append("NO-LABEL")
+
+        span_list_field: ListField = ListField(spans)
+        fields["spans"] = span_list_field
+        if gold_tree is not None:
+            fields["span_labels"] = SequenceLabelField(gold_labels, span_list_field)
+
+        return Instance(fields)
+
+    def _get_gold_spans(self, # pylint: disable=arguments-differ
+                        tree: Tree,
+                        index: int,
+                        typed_spans: OrderedDict) -> OrderedDict:
+        """
+        Recursively construct the gold spans from an nltk ``Tree``.
+        Spans are inclusive.
+
+        Parameters
+        ----------
+        tree : ``Tree``, required.
+            An NLTK parse tree to extract spans from.
+        index : ``int``, required.
+            The index of the current span in the sentence being considered.
+        typed_spans : ``OrderedDict[Tuple[int, int], str]``, required.
+            A ordered dictionary mapping spans to span labels.
+
+        Returns
+        -------
+        typed_spans : ``OrderedDict[Tuple[int, int], str]``, required.
+            A ordered dictionary mapping all subtree spans in the parse tree
+            to their constituency labels. Leaf nodes have POS tag spans, which
+            are denoted by a label of "LABEL-POS".
+        """
+        # NLTK leaves are strings.
+        if isinstance(tree[0], str):
+            # The "length" of a tree is defined by
+            # NLTK as the number of children.
+            # We don't actually want the spans for leaves, because
+            # their labels are pos tags. However, it makes the
+            # indexing more straightforward, so we'll collect them
+            # and filter them out below. We subtract 1 from the end
+            # index so the spans are inclusive.
```

Comment:
Why do you want the spans to be inclusive?

Reply:
because it makes selecting indices from tensors easier if the spans we generate can just be mapped to indices into rows of tensors, rather than having to shift the end index first

Reply:
That could make sense, but depending on if you add START and END tokens, you might have to shift the end index if you don't subtract 1.

Reply:
Yeah this span stuff is quite new, it would be great to get it working easily with just a flag to switch the inclusive exclusive thing. Although if you add start tokens don't you then have to shift things for the start index anyway?

Reply:
You wouldn't have to shift the start index, but that is because we want to index to the word right before the span. But we will have to do some index foo when dealing with the output of the backward lstm.